src/interface/swarm.cpp

```diff
@@ -317,28 +354,469 @@ void Swarm::Defrag() {
+        }
+      });
+
+  // Count all the particles that are Active and Not on this block, if nonzero,
+  // copy into buffers (if no send already for that buffer) and send
+
+  RemoveMarkedParticles();
+
+  vbswarm->Send(phase);
+  return true;
+}
+
+vpack_types::SwarmVarList<Real> Swarm::MakeRealList_(std::vector<std::string> &names) {
+  int size = 0;
+  vpack_types::SwarmVarList<Real> vars;
+
+  for (auto it = realVector_.rbegin(); it != realVector_.rend(); ++it) {
+    auto v = *it;
+    vars.push_front(v);
+    size++;
+  }
+  // Get names in same order as list
+  names.resize(size);
+  int it = 0;
+  for (auto &v : vars) {
+    names[it++] = v->label();
+  }
+  return vars;
+}
+
+vpack_types::SwarmVarList<int> Swarm::MakeIntList_(std::vector<std::string> &names) {
+  int size = 0;
+  vpack_types::SwarmVarList<int> vars;
+
+  for (auto it = intVector_.rbegin(); it != intVector_.rend(); ++it) {
+    auto v = *it;
+    vars.push_front(v);
+    size++;
+  }
+  // Get names in same order as list
+  names.resize(size);
+  int it = 0;
+  for (auto &v : vars) {
+    names[it++] = v->label();
+  }
+  return vars;
+}
+
+SwarmVariablePack<Real> Swarm::PackAllVariablesReal(PackIndexMap &vmap) {
+  std::vector<std::string> names;
+  for (auto &v : realVector_) {
+    names.push_back(v->label());
+  }
+  return PackVariablesReal(names, vmap);
+}
+
+SwarmVariablePack<Real> Swarm::PackVariablesReal(const std::vector<std::string> &names,
+                                                 PackIndexMap &vmap) {
+  std::vector<std::string> expanded_names = names;
+  vpack_types::SwarmVarList<Real> vars = MakeRealList_(expanded_names);
+
+  auto pack = MakeSwarmPack<Real>(vars, &vmap);
+  SwarmPackIndxPair<Real> value;
+  value.pack = pack;
+  value.map = vmap;
+  return pack;
+}
+SwarmVariablePack<int> Swarm::PackVariablesInt(const std::vector<std::string> &names,
+                                               PackIndexMap &vmap) {
+  std::vector<std::string> expanded_names = names;
+  vpack_types::SwarmVarList<int> vars = MakeIntList_(expanded_names);
+
+  auto pack = MakeSwarmPack<int>(vars, &vmap);
+  SwarmPackIndxPair<int> value;
+  value.pack = pack;
+  value.map = vmap;
+  return pack;
+}
+
+bool Swarm::Receive(BoundaryCommSubset phase) {
+  // Ensure all local deep copies marked BoundaryStatus::completed are actually received
+  GetBlockPointer()->exec_space.fence();
+  auto pmb = GetBlockPointer();
+
+  // Populate buffers
+  vbswarm->Receive(phase);
+
+  // Copy buffers into swarm data on this proc
+  int maxneighbor = vbswarm->bd_var_.nbmax;
+  int total_received_particles = 0;
+  std::vector<int> neighbor_received_particles(maxneighbor);
+  for (int n = 0; n < maxneighbor; n++) {
+    if (vbswarm->bd_var_.flag[pmb->pbval->neighbor[n].bufid] == BoundaryStatus::arrived) {
+      total_received_particles += vbswarm->recv_size[n] / vbswarm->particle_size;
+      neighbor_received_particles[n] = vbswarm->recv_size[n] / vbswarm->particle_size;
```

Comment:
Integer division can be risky, are there any checks in place to be sure you are getting the right result?

Reply:
I added a ` PARTHENON_DEBUG_REQUIRE(vbswarm->recv_size[n] % vbswarm->particle_size == 0, "Receive buffer is not divisible by particle size!");`, is that what you had in mind?

Also I swapped those two lines so we only actually evaluate the division once.